### PR TITLE
Introduce XML writing operations for DOCTYPE declaration writing

### PR DESCRIPTION
**Pull request by BuddhiWathsala**
## Purpose
Fixes https://github.com/ballerina-platform/ballerina-standard-library/issues/941

## Approach
Introduce XML writing options to the XML write API

- FileWriteOption - used to indicate whether to overwrite or append the content. Append only applicable when writing external parsed entities.
- XmlWriteOptions - to indicate XML doc type and entity type. The `xmlEntityType` to set the type of the document(document or external parsed entity).
- XmlDoctype - if an internal subset is specified, it is given priority over external ones and create the DOCTYPE declaration. If only the system is specified, it created a DOCTYPE declaration with SYSTEM. If public or (public and system) both specified, then it creates a DOCTYPE with PUBLIC.

## Automation tests
 - Unit tests 
   > Yes

### Sample with SYSTEM
```ballerina 
import ballerina/io;

public function main() returns error? {
    xml content = xml `<note>
        <to>Tove</to>
        <from>Jani</from>
        <heading>Reminder</heading>
        <body>Don't forget me this weekend!</body>
        </note>`;
    check io:fileWriteXml("./sample.xml", content, doctype = {system: "Note.dtd"});
}
```

Output
```xml
<!DOCTYPE note SYSTEM "Note.dtd">
<note>
    <to>Tove</to>
    <from>Jani</from>
    <heading>Reminder</heading>
    <body>Don't forget me this weekend!</body>
</note>
```

### Sample with PUBLIC
```ballerina
import ballerina/io;

public function main() returns error? {
    xml content = xml `<note>
        <to>Tove</to>
        <from>Jani</from>
        <heading>Reminder</heading>
        <body>Don't forget me this weekend!</body>
        </note>`;
    string publicId = "-//W3C//DTD HTML 4.01 Transitional//EN";
    string systemId = "http://www.w3.org/TR/html4/loose.dtd";
    var writeResult = io:fileWriteXml("./sample.xml", content, doctype = {
        system: systemId,
        'public: publicId
    });
}
```

Output
```xml
<!DOCTYPE note PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<note>
    <to>Tove</to>
    <from>Jani</from>
    <heading>Reminder</heading>
    <body>Don't forget me this weekend!</body>
</note>
```

### Sample with Internal Declaration
```ballerina
import ballerina/io;

public function main() returns error? {
    xml content = xml `<note>
        <to>Tove</to>
        <from>Jani</from>
        <heading>Reminder</heading>
        <body>Don't forget me this weekend!</body>
        </note>`;
    string internalSub = string `[
        <!ELEMENT note (to,from,heading,body)>
        <!ELEMENT to (#PCDATA)>
        <!ELEMENT from (#PCDATA)>
        <!ELEMENT heading (#PCDATA)>
        <!ELEMENT body (#PCDATA)>
    ]`;
    var writeResult = io:fileWriteXml(filePath, content, doctype = {
        internalSubset: internalSub,
        system: systemId
    });
}
```

Output
```xml
<!DOCTYPE note [
        <!ELEMENT note (to,from,heading,body)>
        <!ELEMENT to (#PCDATA)>
        <!ELEMENT from (#PCDATA)>
        <!ELEMENT heading (#PCDATA)>
        <!ELEMENT body (#PCDATA)>
    ]>
<note>
    <to>Tove</to>
    <from>Jani</from>
    <heading>Reminder</heading>
    <body>Don't forget me this weekend!</body>
</note>
```

## Test environment
- macOS
- SLAlpha 3
- Java11
 
## Learning
- https://www.w3.org/TR/xml/#NT-doctypedecl
- https://www.liquid-technologies.com/XML/DocType.aspx

### CODE REVIEW

File: io-ballerina/file_string_io.bal [hunk 1]
```diff
@@ -173,12 +173,66 @@ public function fileWriteJson(@untainted string path, json content) returns @tai
 # ```
 # + path - The path of the XML file
 # + content - XML content to write
+# + xmlOptions - XML writing options(XML entity type and DOCTYPE)
+# + fileWriteOption - file write option(`OVERWRITE` and `APPEND` are the possible values, and the default value is `OVERWRITE`)
 # + return - The null `()` value when the writing was successful or an `io:Error`
-public function fileWriteXml(@untainted string path, xml content) returns Error? {
-    var byteChannel = openWritableFile(path);
+public function fileWriteXml(@untainted string path, xml content, *XmlWriteOptions xmlOptions, FileWriteOption fileWriteOption=OVERWRITE) returns Error? {
```
Comment (daneshk) [inline io-ballerina/file_string_io.bal:179]:
```suggestion
public function fileWriteXml(@untainted string path, xml content, *XmlWriteOptions xmlOptions, FileWriteOption fileWriteOption = OVERWRITE) returns Error? {
```

File: io-ballerina/options.bal [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+# Represents a file opening options for writing.
+#
+# + OVERWRITE - Overriwrite(truncate the existing content)
```
Comment (daneshk) [inline io-ballerina/options.bal:19]:
```suggestion
# + OVERWRITE - Overwrite(truncate the existing content)
```

File: io-ballerina/file_string_io.bal [hunk 3]
```diff
@@ -173,12 +173,66 @@ public function fileWriteJson(@untainted string path, json content) returns @tai
...
+# + fileWriteOption - file write option(`OVERWRITE` and `APPEND` are the possible values, and the default value is `OVERWRITE`)
 # + return - The null `()` value when the writing was successful or an `io:Error`
-public function fileWriteXml(@untainted string path, xml content) returns Error? {
-    var byteChannel = openWritableFile(path);
+public function fileWriteXml(@untainted string path, xml content, *XmlWriteOptions xmlOptions, FileWriteOption fileWriteOption = OVERWRITE) returns Error? {
+    WritableByteChannel|Error byteChannel;
+    xml writeContent = xml ``;
+    if (xmlOptions.xmlEntityType == DOCUMENT_ENTITY) {
+        if (fileWriteOption == APPEND) {
+            return error ConfigurationError("The APPEND operation not allowed with DOCUMENT");
```
Comment (daneshk) [inline io-ballerina/file_string_io.bal:184]:
```suggestion
            return error ConfigurationError("The file append operation is not allowed for Document Entity");
```

File: io-ballerina/file_string_io.bal [hunk 4]
```diff
@@ -173,12 +173,66 @@ public function fileWriteJson(@untainted string path, json content) returns @tai
...
+        }
+        if (xml:length(content) > 1) {
+            return error ConfigurationError("The DOCUMENT XML can only contains single root");
```
Comment (daneshk) [inline io-ballerina/file_string_io.bal:187]:
```suggestion
            return error ConfigurationError("The XML Document can only contain single root");
```

File: io-ballerina/file_string_io.bal [hunk 5]
```diff
@@ -173,12 +173,66 @@ public function fileWriteJson(@untainted string path, json content) returns @tai
...
     } else {
         return byteChannel;
     }
 }
+
+function populateDoctype(xml content, XmlDoctype doctype) returns string {
+    // Generate <!DOCTYPE rootElementName PUBLIC|SYSTEM PublicIdentifier SystemIdentifier internalSubset>
+    string doctypeElement = "";
+    string startElement = "<!DOCTYPE ";
+    string endElement = ">";
```
Comment (daneshk) [inline io-ballerina/file_string_io.bal:215]: Shall we use a string template for this

Comment (daneshk) [reply io-ballerina/file_string_io.bal:215]: https://ballerina.io/learn/by-example/string-template.html

Comment (BuddhiWathsala) [reply io-ballerina/file_string_io.bal]: Fixed in https://github.com/ballerina-platform/module-ballerina-io/pull/77/commits/c0d902a6e341c5c475b6834b65f5ab7eb5a82b07

File: io-ballerina/options.bal (hunk 2 above)
Comment (BuddhiWathsala) [reply io-ballerina/options.bal]: Fixed in https://github.com/ballerina-platform/module-ballerina-io/pull/77/commits/96947ca73cc598ad57975b7931a4ebff0ca199ff

File: io-ballerina/file_string_io.bal (hunk 3 above)
Comment (BuddhiWathsala) [reply io-ballerina/file_string_io.bal]: Fixed in https://github.com/ballerina-platform/module-ballerina-io/pull/77/commits/96947ca73cc598ad57975b7931a4ebff0ca199ff

File: io-ballerina/file_string_io.bal (hunk 4 above)
Comment (BuddhiWathsala) [reply io-ballerina/file_string_io.bal]: Fixed in https://github.com/ballerina-platform/module-ballerina-io/pull/77/commits/96947ca73cc598ad57975b7931a4ebff0ca199ff

**Review by daneshk — APPROVED** [review]

**Pull request merged**